src/control/family.lean

```diff
@@ -0,0 +1,666 @@
+
+/-- predicate over the values in type family `α` -/
+@[pp_nodot, derive inhabited]
+def Pred (α : fam I) : Sort* := ∀ i, α i → Prop
+
+/-- introduction rule for `Pred α` -/
+def Pred.mk {α : fam I} (p : Π i, (unit i ⟶ α) → Prop) : Pred α :=
+λ i x, p i $ value i _ x
+
+/-- elimination rule for `Pred α` -/
+def Pred.apply {α : fam I} (p : Pred α) : ∀ ⦃i⦄, (unit i ⟶ α) → Prop :=
+λ i f, p i $ f unit.rfl
+
+@[simp]
+lemma Pred.apply_mk {α : fam I} (p : Π i, (unit i ⟶ α) → Prop) :
+  Pred.apply (Pred.mk p) = p :=
+by ext : 2; simp [Pred.apply,Pred.mk]; congr'; ext _ ⟨ ⟩; refl
+
+@[simp]
+lemma Pred.mk_to_fun {α : fam I} (p : Π i, (unit i ⟶ α) → Prop) {i} (x : α i) :
+  Pred.mk p i x = p i (value i _ x) := rfl
+
+@[simp]
+lemma Pred.mk_apply {α : fam I} (p : Pred α) :
+  Pred.mk (Pred.apply p) = p := by ext; refl
+
+/-- contravariant map function for `Pred` -/
+def Pred.map {α β : fam I} (f : α ⟶ β) (p : Pred β) : Pred α :=
+λ i x, p i (f x)
+
+lemma Pred.map_mk {α β : fam I} (f : α ⟶ β) (p : Π ⦃i⦄, (unit i ⟶ β) → Prop) :
+  Pred.map f (Pred.mk p) = Pred.mk (λ i g, p (g ≫ f)) :=
+by ext; simp [Pred.mk,Pred.map]; congr'; ext _ ⟨ ⟩; refl
+
+/-- subtypes as an object of a type family category -/
+@[reducible]
+def subtype {α : fam I} (p : Pred α) : fam I :=
+λ i, subtype (p i)
+
+/-- elimination rule for `subtype` object -/
+def subtype.val {α : fam I} {p : Pred α} : fam.subtype p ⟶ α :=
+λ i, subtype.val
+
+/-- map function on the predicate of `subtype` -/
+def subtype.map {α β : fam I} (p : Pred α) (q : Pred β)
+  (f : α ⟶ β) (h : ∀ i x, p i x → q i (f x)) :
+  fam.subtype p ⟶ fam.subtype q :=
+λ i (x : subtype p i), subtype.mk (f x.1) (h i _ x.2)
+
+lemma subtype.ext {α : fam I} {p : Pred α } {X} (a b : X ⟶ subtype p)
+  (h : a ≫ subtype.val = b ≫ subtype.val) : a = b :=
+by ext : 2; rw subtype.ext_iff; apply congr_fun (congr_fun h x)
+
+lemma subtype.map_val {α β : fam I} {p : Pred α} {q : Pred β} (a : α ⟶ β) (h) :
+  subtype.map p q a h ≫ subtype.val = subtype.val ≫ a :=
+by ext _ ⟨ ⟩ : 2; refl
+
+/-- binary product in the category `fam I` -/
+def prod (α β : fam I) : fam I
+| i := α i × β i
+
+localized "infix ` ⊗ `:35 := fam.prod" in fam
+
+/-- left projection of binary product in the category `fam I` -/
+def prod.fst : Π {α β : fam I}, α ⊗ β ⟶ α
+| α β i x := _root_.prod.fst x
+
+/-- right projection of binary product in the category `fam I` -/
+def prod.snd : Π {α β : fam I}, α ⊗ β ⟶ β
+| α β i x := _root_.prod.snd x
+
+/-- map function of the binary product in the category `fam I` -/
+def prod.map {α β α' β' : fam I} : (α ⟶ β) → (α' ⟶ β') → (α ⊗ α' ⟶ β ⊗ β')
+| f g i x := (f x.1,g x.2)
+
+localized "infix ` ⊗' `:35 := fam.prod.map" in fam
+
+@[simp, reassoc]
+lemma prod.map_fst {α β α' β' : fam I} (f : α ⟶ β) (g : α' ⟶ β') :
+  prod.map f g ≫ prod.fst = prod.fst ≫ f :=
+by ext; refl
+
+@[simp, reassoc]
+lemma prod.map_snd {α β α' β' : fam I} (f : α ⟶ β) (g : α' ⟶ β') :
+  prod.map f g ≫ prod.snd = prod.snd ≫ g :=
+by ext; refl
```

Comment:
We could remove all these sorts of lemmas by "just" providing the instance `has_binary_products (fam I)`. Then we'd get them for free.

What do you think? I can have a go at this if it would be helpful.

Reply:
That'd be great! Thanks! I was a bit worried about reusing the product infrastructure on `fam`. I think the definition product categories would be problematic for this development but tensor product / coproduct should work.